// winrow/lib/server.js
'use strict';

const Promise = require('winrow/lib/node_modules/bluebird');
const lodash = require('winrow/lib/node_modules/lodash');
const http = require('http');
const path = require('path');
const express = require('winrow/lib/node_modules/express');
const app = express();
const router = express.Router();
const logger = require('winrow/lib/node_modules/morgan');
const mongoose = require('mongoose');
const cors = require('winrow/lib/node_modules/cors');
const bodyParser = require('winrow/lib/node_modules/body-parser');
const moment = require('winrow/lib/node_modules/moment');
const momentTimezone = require('moment-timezone');
const slugifyString = require('winrow/config/utils/slugifyString');
const returnCodes = require('winrow/config/utils/returnCodes');
const validator = require('winrow/config/utils/validator');
const getRequestId = require('winrow/config/utils/getRequestId');
const loggingFactory = require('winrow/lib/node_modules/app-logger');
const { isFunction } = lodash;

function Server() {
  // create server
  const server = http.createServer(app);
  // get Promise
  this.Promise = Promise;
  // get Lodash
  this.lodash = lodash;
  // get mongoose
  this.mongoose = mongoose;
  // global Promise
  mongoose.Promise = global.Promise;
  mongoose.set("debug", true);
  // get moment
  this.moment = moment;
  this.momentTimezone = momentTimezone;
  // get slug
  this.slugifyString = slugifyString;
  // get returnCodes
  this.returnCodes = returnCodes;
  // get validator
  this.validator = validator;
  // get requestId
  this.getRequestId = getRequestId;
  const requestId = getRequestId();
  //get app
  this.app = app;
  //get router
  this.router = router;
  // get express
  this.express = express;
  app.use(logger('dev'));
  // connect ui
  app.use(cors());
  app.use(bodyParser.json({ limit: '100mb' }));
  app.use(bodyParser.urlencoded({ limit: '100mb', extended: true }));
  app.use(express.static(path.join(__dirname, 'build')));
  app.use('/rest/api/images', express.static('images'));
  // start server
  this.start = function () {
    const protocolHttp = process.env.PROTOCOL || 'http';
    const portHttp = process.env.PORT || 8080;
    const hostHttp = process.env.HOST || '0.0.0.0';
    return new Promise((resolve, rejects) => {
      const serverInstance = server.listen(portHttp, hostHttp, () => {
        const port = serverInstance.address().port;
        const host = serverInstance.address().address;
        console.info('The server http is running on %s://%s:%s', protocolHttp, host, port);
      })
      resolve(serverInstance)
    })
      .then(info => {
        loggingFactory.warn('Start server', { requestId: `${requestId}` })
        return info
      })
      .catch(err => {
        return Promise.reject(err);
      })
  };
  // stop server
  this.stop = function () {
    return new Promise(function (resolve, reject) {
      const timeOut = setTimeout(function () {
        reject();
      }, 3000);
      loggingFactory.warn("time out", timeOut, { requestId: `${requestId}` });
      const serverClose = function () {
        if (server && isFunction(serverClose)) {
          server.removeListener('close', serverClose)
        }
      }
      loggingFactory.warn('The Server Close');
      server.on('close', serverClose)
      server.close(function (err) {
        clearTimeout(timeOut)
        resolve();
      });
    })
      .then(() => {
        console.info('The Server Has Been Closed');
        process.exit(1);
      })
      .catch(err => {
        return Promise.reject(err)
      })
  };
};

module.exports = new Server();
